client/consul.go

```diff
@@ -0,0 +1,159 @@
+
+// NewConsulClient constructs a consul api client
+func NewConsulClient(conf *config.Config, maxRetry int) (*ConsulClient, error) {
+	consulConf := conf.Consul
+	transport := hcat.TransportInput{
+		SSLEnabled: *consulConf.TLS.Enabled,
+		SSLVerify:  *consulConf.TLS.Verify,
+		SSLCert:    *consulConf.TLS.Cert,
+		SSLKey:     *consulConf.TLS.Key,
+		SSLCACert:  *consulConf.TLS.CACert,
+		SSLCAPath:  *consulConf.TLS.CAPath,
+		ServerName: *consulConf.TLS.ServerName,
+
+		DialKeepAlive:       *consulConf.Transport.DialKeepAlive,
+		DialTimeout:         *consulConf.Transport.DialTimeout,
+		DisableKeepAlives:   *consulConf.Transport.DisableKeepAlives,
+		IdleConnTimeout:     *consulConf.Transport.IdleConnTimeout,
+		MaxIdleConns:        *consulConf.Transport.MaxIdleConns,
+		MaxIdleConnsPerHost: *consulConf.Transport.MaxIdleConnsPerHost,
+		TLSHandshakeTimeout: *consulConf.Transport.TLSHandshakeTimeout,
+	}
+
+	consul := hcat.ConsulInput{
+		Address:      *consulConf.Address,
+		Token:        *consulConf.Token,
+		AuthEnabled:  *consulConf.Auth.Enabled,
+		AuthUsername: *consulConf.Auth.Username,
+		AuthPassword: *consulConf.Auth.Password,
+		Transport:    transport,
+	}
+
+	clients := hcat.NewClientSet()
+	if err := clients.AddConsul(consul); err != nil {
+		return nil, err
+	}
+
+	logger := logging.Global().Named(loggingSystemName).Named(consulSubsystemName)
+
+	r := retry.NewRetry(maxRetry, time.Now().UnixNano())
+	return &ConsulClient{
+		Client: clients.Consul(),
+		retry:  r,
+		logger: logger}, nil
+}
+
+// GetLicense queries Consul for a signed license, and returns it if available
+func (c *ConsulClient) GetLicense(ctx context.Context, q *consulapi.QueryOptions) (string, error) {
+	c.logger.Debug("getting license")
+
+	desc := "consul client get license"
+	var err error
+	var license string
+
+	err = c.retry.Do(ctx, func(context.Context) error {
+		license, err = c.Operator().LicenseGetSigned(q)
+		if err != nil {
+			license = ""
+			return err
+		}
+		return nil
+	}, desc)
+
+	return license, err
+}
+
+// GetSKU queries Consul for information about itself, it then
+// parses this information to determine whether the Consul being
+// queried is Enterprise or OSS
+func (c *ConsulClient) GetSKU(ctx context.Context) (string, error) {
+	c.logger.Debug("getting sku")
+
+	desc := "consul client get sku"
+	var err error
+	var info Self
+
+	err = c.retry.Do(ctx, func(context.Context) error {
+		info, err = c.Agent().Self()
+		if err != nil {
+			info = nil
+			return err
+		}
+		return nil
+	}, desc)
+	if err != nil {
+		return "", err
+	}
+
+	sku, ok := parseSKU(info)
+	if !ok {
+		return "", errors.New("unable to parse sku")
+	}
+	return sku, nil
+}
+
+func parseSKU(info Self) (string, bool) {
+	v, ok := info["Config"]["Version"].(string)
+	if !ok {
+		return "", ok
+	}
+
+	ver, vErr := version.NewVersion(v)
+	if vErr != nil {
+		return "", false
+	}
+	if strings.Contains(ver.Metadata(), ConsulEnterpriseSKU) {
```

Comment:
Nit, space out code blocks for better readability. The second `if` can be confused with an `else` based on how the code is spaced.